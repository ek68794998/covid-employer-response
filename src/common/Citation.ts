import { CitationSource } from "./CitationSource";
import { CitationType } from "./CitationType";

export class Citation {
	public positivity: number = 0;

	public sources?: CitationSource[];

	public summary: string = "";

	public type?: CitationType;
}
