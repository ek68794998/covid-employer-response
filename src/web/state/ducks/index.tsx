import { combineReducers } from "redux";

import { getLocalizedStrings } from "./localization/reducers";

export default combineReducers({
	strings: getLocalizedStrings,
});