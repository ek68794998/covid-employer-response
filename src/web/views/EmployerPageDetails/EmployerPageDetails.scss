@import "../Dimensions.scss";

$padding-interior: 18px;

div.EmployerPageDetails__Container {
	font-size: 16px;

	a.EmployerActionLinks__Link {
		opacity: 1.0 !important;
	}
}

div.EmployerPageDetails__Body > div {
	margin-top: $padding-interior;

	&.EmployerPageDetails__Summary {
		margin-top: 6px;
	}
}

div.EmployerPageDetails__Summary {
	line-height: 180%;
}
