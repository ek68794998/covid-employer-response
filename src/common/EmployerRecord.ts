import { Citation } from "./Citation";
import { EmployerLocation } from "./EmployerLocation";
import { EmployerRating } from "./EmployerRating";

export class EmployerRecord {
	public citations: Citation[] = [];

	public employeesBeforeMax: number = -1;

	public employeesBeforeMin: number = -1;

	public id: string = "";

	public image?: string;

	public location: EmployerLocation = new EmployerLocation();

	public name: string = "";

	public officialWebsite?: string;

	public rating: EmployerRating = "fair";

	public summary: string = "";

	public wiki?: string;
}
