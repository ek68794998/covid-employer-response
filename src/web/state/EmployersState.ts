import { EmployerRecord } from "../../common/EmployerRecord";

export interface EmployersState {
	items: EmployerRecord[];
}