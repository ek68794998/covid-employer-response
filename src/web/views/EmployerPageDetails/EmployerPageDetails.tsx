import React from "react";
import { RouteProps } from "react-router-dom";

import { Citation } from "../../../common/Citation";
import { CitationType } from "../../../common/CitationType";
import { EmployerRecord } from "../../../common/EmployerRecord";

import EmployerCitationList from "../EmployerCitationList/EmployerCitationList";
import EmployerDetailsHeader from "../EmployerDetailsHeader/EmployerDetailsHeader";

import "./EmployerPageDetails.scss";

interface Props extends RouteProps {
	employer: EmployerRecord;
}

const citationSort = (a: Citation, b: Citation): number => {
	if (a.positivity !== b.positivity) {
		return b.positivity - a.positivity;
	}

	return (b.sources ? b.sources.length : 0) - (a.sources ? a.sources.length : 0);
};

const EmployerPageDetails: React.FC<Props> = (props: Props): React.ReactElement | null => {
	const { employer } = props;

	if (!employer) {
		return null;
	}

	let globalCitationSourceBase: number = 1;

	const components: Partial<{ [key in CitationType]: JSX.Element | null }> = {};

	for (const type of [ "publication", "statement", "hearsay" ]) {
		const citations: Citation[] =
			employer.citations
				.filter((citation: Citation) => citation.type === type)
				.sort(citationSort);

		if (citations.length === 0) {
			continue;
		}

		components[type] = (
			<EmployerCitationList
				citations={citations}
				citationSourceBase={globalCitationSourceBase}
				citationType={type as CitationType}
			/>
		);

		citations.forEach((citation: Citation) => globalCitationSourceBase += citation.sources?.length || 0);
	}

	return (
		<div className="EmployerPageDetails__Container">
			<EmployerDetailsHeader employer={EmployerRecord.toMetadata(employer)} />
			<div className="EmployerPageDetails__Body">
				<div className="EmployerPageDetails__Summary">
					{employer.summary}
				</div>
				{components.publication}
				{components.statement}
				{components.hearsay}
			</div>
		</div>
	);
};

export default EmployerPageDetails;
