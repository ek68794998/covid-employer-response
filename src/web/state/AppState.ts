import { EmployersState } from "./EmployersState";

import { LocalizedStrings } from "../../common/LocalizedStrings";

export interface AppState {
	employers: EmployersState;

	environment: string;

	strings: LocalizedStrings;
}