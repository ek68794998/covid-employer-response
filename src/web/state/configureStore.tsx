import { AnyAction, applyMiddleware, compose, createStore, Store } from "redux";
import thunk from "redux-thunk";

import { Production } from "../../common/constants/EnvironmentConstants";

import { AppState } from "./AppState";
import rootReducer from "./ducks";

const configureStore = (preloadedState: Partial<AppState>): Store<AppState, AnyAction> => {
	const isProd: boolean = process.env.NODE_ENV === Production;
	const composeEnhancers: any =
		(!isProd && typeof window !== "undefined" && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

	const store: Store<AppState, AnyAction> = createStore(
		rootReducer,
		preloadedState,
		composeEnhancers(applyMiddleware(thunk)),
	);

	if (module.hot) {
		// Enable Webpack hot module replacement for reducers
		module.hot.accept("./ducks", () => {
			const nextRootReducer: any = require("./ducks").default;
			store.replaceReducer(nextRootReducer);
		});
	}

	return store;
};

export default configureStore;