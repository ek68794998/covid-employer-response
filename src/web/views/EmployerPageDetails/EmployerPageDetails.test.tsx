import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import { AnyAction, Store } from "redux";

import { mockComponent, ploc } from "../../../__tests__/TestUtils";
import { EmployerRecord } from "../../../common/EmployerRecord";

import { AppState } from "../../state/AppState";
import configureStore from "../../state/configureStore";

import EmployerPageDetails from "./EmployerPageDetails";

jest.mock(
	"../EmployerDetailsHeader/EmployerDetailsHeader",
	() => mockComponent("EmployerDetailsHeader"));

jest.mock(
	"../EmployerCitationList/EmployerCitationList",
	() => mockComponent("EmployerCitationList"));

jest.mock(
	"../EmployerListItem/EmployerListItem",
	() => mockComponent("EmployerListItem"));

describe("<EmployerPageDetails />", () => {
	const createConfigStore = (preloaded: Partial<AppState> = {}): Store<AppState, AnyAction> => configureStore({
		strings: {
			detailDescriptions: {
				aka: ploc("aka"),
				edit: ploc("edit"),
				employees: ploc("employees"),
				industry: ploc("industry"),
				linkToEmployer: ploc("linkToEmployer"),
				location: ploc("location"),
				name: ploc("name"),
				officialWebsite: ploc("officialWebsite"),
				rating: ploc("rating"),
				ratingCounts: ploc("ratingCounts"),
				ticker: ploc("ticker"),
				wikipedia: ploc("wikipedia"),
			},
			detailLabels: {
				officialWebsite: ploc("officialWebsiteLabel"),
				wikipedia: ploc("wikipediaLabel"),
			},
			employeeDelta: `${ploc("employeeDelta")}: {change} / {date}`,
			ratingLabels: {
				fair: ploc("fair"),
				good: ploc("good"),
				poor: ploc("poor"),
			},
		},
		...preloaded,
	});

	test("renders without exploding", () => {
		const e: EmployerRecord = new EmployerRecord();

		const store: Store<AppState, AnyAction> = createConfigStore({
			employers: {
				itemsComplete: {
					[e.id]: e,
				},
				itemsMetadata: {
					[e.id]: EmployerRecord.toMetadata(e),
				},
			},
		});

		const renderedValue: ReactTestRendererJSON | null =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<EmployerPageDetails employer={new EmployerRecord()} />
					</BrowserRouter>
				</Provider>,
			).toJSON();

		expect(renderedValue).toMatchSnapshot();
	});

	test("renders using an employer record", () => {
		const e: EmployerRecord =
			Object.assign(new EmployerRecord(), {
				citations: [{
					positivity: 2,
					summary: "A very good summary about Contoso.",
					type: "hearsay",
				}, {
					positivity: 0,
					sources: [ { name: "CBS", link: "http://cbs", date: "2015-01-01T01:01:01Z" } ],
					summary: "A very good summary about Contoso.",
					type: "publication",
				}, {
					positivity: -1,
					sources: [ { name: "NBC", link: "http://nbc", date: "2016-01-01T01:01:01Z" } ],
					summary: "A very good summary about Contoso.",
					type: "statement",
				}],
				employeesBefore: { type: "exactly", upperBound: 12345, year: 2015 },
				id: "contoso",
				image: "consoto.svg",
				location: { city: "City", country: "us", international: true },
				name: "Contoso",
				summary: "Contoso is a good company.",
			});

		const store: Store<AppState, AnyAction> = createConfigStore({
			employers: {
				itemsComplete: {
					[e.id]: e,
				},
				itemsMetadata: {
					[e.id]: EmployerRecord.toMetadata(e),
				},
			},
		});

		const renderedValue: ReactTestRendererJSON | null =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<EmployerPageDetails employer={e} />
					</BrowserRouter>
				</Provider>,
			).toJSON();

		expect(renderedValue).toMatchSnapshot();
	});
});
