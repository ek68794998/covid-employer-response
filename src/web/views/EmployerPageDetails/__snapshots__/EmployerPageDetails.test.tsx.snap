// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmployerPageDetails /> renders using an employer record 1`] = `
<div
  className="EmployerPageDetails__Container"
>
  <div
    className="EmployerPageDetails__Header"
  >
    <img
      className="EmployerLogo__Image"
      src="/images/employers/consoto.svg"
      style={
        Object {
          "background": "#fff",
        }
      }
    />
    <h1
      className="EmployerPageDetails__Title"
    >
      Contoso
    </h1>
    <div
      className="EmployerPageDetails__Rating EmployerPageDetails__Rating--fair"
    >
      ƒáïř
      <i
        className="material-icons"
      >
        trending_flat
      </i>
    </div>
  </div>
  <div
    className="EmployerPageDetails__Summary"
  >
    <p>
      Contoso is a good company.
    </p>
  </div>
  [component: EmployerCitationList citations=([object Object],[object Object],[object Object]) /]
  <div
    className="EmployerPageDetails__MetaFooter"
  >
    <div
      className="EmployerPageDetails__Profile"
    >
      <h2 />
      <table
        className="EmployerPageDetails__ProfileDetails"
      >
        <tbody>
          <tr>
            <td
              className="EmployerPageDetails__ProfileLabel"
              title="ñá₥è"
            >
              <i
                className="material-icons"
              >
                work
              </i>
            </td>
            <td
              className="EmployerPageDetails__ProfileValue"
            >
              Contoso
            </td>
          </tr>
          <tr>
            <td
              className="EmployerPageDetails__ProfileLabel"
              title="ℓôçáƭïôñ"
            >
              <i
                className="material-icons"
              >
                place
              </i>
            </td>
            <td
              className="EmployerPageDetails__ProfileValue"
            >
              City, us (Multinational)
              <img
                className="EmployerPageDetails__ProfileFlag"
                src="/images/flags/us.svg"
              />
            </td>
          </tr>
          <tr>
            <td
              className="EmployerPageDetails__ProfileLabel"
              title="è₥ƥℓô¥èèƨ"
            >
              <i
                className="material-icons"
              >
                people
              </i>
            </td>
            <td
              className="EmployerPageDetails__ProfileValue"
            >
              ~10,123 (2017)
              <span
                className="EmployerPageDetails__Delta EmployerPageDetails__Delta--Negative"
              >
                (
                è₥ƥℓô¥èèÐèℓƭá: -2,222 / 2015
                )
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      className="EmployerPageDetails__LinkedCompanies"
    />
  </div>
</div>
`;

exports[`<EmployerPageDetails /> renders without exploding 1`] = `
<div
  className="EmployerPageDetails__Container"
>
  <div
    className="EmployerPageDetails__Header"
  >
    <h1
      className="EmployerPageDetails__Title"
    >
      
    </h1>
    <div
      className="EmployerPageDetails__Rating EmployerPageDetails__Rating--fair"
    >
      ƒáïř
      <i
        className="material-icons"
      >
        trending_flat
      </i>
    </div>
  </div>
  <div
    className="EmployerPageDetails__Summary"
  />
  [component: EmployerCitationList citations=() /]
  <div
    className="EmployerPageDetails__MetaFooter"
  >
    <div
      className="EmployerPageDetails__Profile"
    >
      <h2 />
      <table
        className="EmployerPageDetails__ProfileDetails"
      >
        <tbody>
          <tr>
            <td
              className="EmployerPageDetails__ProfileLabel"
              title="ñá₥è"
            >
              <i
                className="material-icons"
              >
                work
              </i>
            </td>
            <td
              className="EmployerPageDetails__ProfileValue"
            >
              
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      className="EmployerPageDetails__LinkedCompanies"
    />
  </div>
</div>
`;
